linkifiers: Implement URI Template in replacement of URL format strings.

PIG208 opened the pull request:
Fixes: #23124

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

andersk commented on zerver/migrations/0422_backfill_realmfilter_url_template.py:
```diff
@@ -0,0 +1,42 @@
+
+    linkifiers = linkifier_model.objects.all()
+    # Matches anything of the form %(variable_name)
+    var_pattern = re.compile(r"%\(([a-zA-Z0-9_-]+)\)s")
+    for linkifier in linkifiers:
+        # Replace format string variables with the RFC 6570 URI Template syntax
+        converted_template = var_pattern.sub(r"{\1}", linkifier.url_format)
```
This doesn’t correctly handle patterns containing `%%`, which is an escaped `%` in the language of format strings. We need to convert

`http://foo/%%` → `http://foo/%`
`http://foo/%%%%` → `http://foo/%%`
`http://foo/%%%%%%` → `http://foo/%%%`

`http://foo/%(var)s` → `http://foo/{var}`
`http://foo/%%(var)s` → `http://foo/%(var)s`
`http://foo/%%%(var)s` → `http://foo/%{var}`
`http://foo/%%%%(var)s` → `http://foo/%%(var)s`
`http://foo/%%%%%(var)s` → `http://foo/%%{var}`
`http://foo/%%%%%%(var)s` → `http://foo/%%%(var)s`

andersk replied:
We also need to quote any pre-existing `{` and `}` as `%7B` and `%7D`.

andersk commented on zerver/migrations/0422_backfill_realmfilter_url_template.py:
```diff
@@ -0,0 +1,49 @@
+    )
+    for linkifier in linkifiers:
+        # Replace format string variables with the RFC 6570 URI Template syntax
+        converted_template = var_pattern.sub(r"\1{\2}", linkifier.url_format)
+        converted_template = converted_template.replace("%%", "%").translate(escape_table)
```
The translation from `{}` to `%7B%7D` needs to be before the translation from `%()s` to `{}`.

andersk commented on zerver/migrations/0422_backfill_realmfilter_url_template.py:
```diff
@@ -0,0 +1,49 @@
+        }
+    )
+    for linkifier in linkifiers:
+        converted_template = linkifier.url_format.replace("%%", "%").translate(escape_table)
+        # Replace format string variables with the RFC 6570 URI Template syntax
+        converted_template = var_pattern.sub(r"\1{\2}", converted_template)
```
This will still fail on `http://foo/%%(var)s` (will give `http://foo/{var}`, expected `http://foo/%(var)s`, because that’s the whole point of escaping).

PIG208 replied:
Oh, `.replace("%%", "%")` goes after `.sub()`.

PIG208 commented:
This is ready for review @timabbott 
I walked through and cleaned up the commits again, and added a migration test case for the backfilling migration.
We will also need to coordinate this change with python-zulip-api since `url_format_string` gets removed for the `POST/PATCH` endpoints.

andersk commented on static/js/linkifiers.js:
```diff
@@ -17,7 +17,7 @@ function python_to_js_linkifier(pattern, url) {
         pattern = pattern.replace("(?P<" + name + ">", "(");
         // Replace named reference in URL to numbered reference
-        url = url.replace("%(" + name + ")s", "\\" + current_group);
+        url = url.replace("{" + name + "}", "\\" + current_group);
```
URI Templates are more than just `{foo}`; an example from the RFC is `http://example.com/search{?q,lang}`. We want a real RFC 6570 implementation here, such as [url-template](https://www.npmjs.com/package/url-template).

andersk commented on zerver/tests/test_realm_linkifiers.py:
```diff
@@ -210,15 +203,15 @@ def test_valid_urls(self) -> None:
-            "https://example.com/?foo=%(foo)s",
+            "https://example.com/{foo}",
+            "https://example.com/{foo}{bars}",
+            "https://example.com/{foo}/and/{bar}",
+            "https://example.com/?foo={foo}",
```
We might as well throw in some of the nontrivial URI Template features here.

```suggestion
            "https://example.com/{foo}/and{/bar}",
            "https://example.com/{?foo}",
```

PIG208 commented:
I get `url-template` integrated into the frontend, and manually tested it on the dev server. However, I have been running into trouble when trying to set up node tests for it. It seems that [url-template](https://www.npmjs.com/package/url-template) only supports esm and it fails `markdown.js` with `ERR_REQUIRE_ESM`.

PIG208 commented:
I updated the PR to use `url-template@2.0.8` to get around the ESM issue. Now the frontend should have full URL template support. I'm not sure if the puppeteer test failure is relevant here, as it also appears on `main`.

timabbott commented:
The puppeteer failure is a flake; you can rebase this as I've now merged https://github.com/zulip/zulip/pull/23731. Thanks for pushing on this!

PIG208 commented:
I have rebased the PR to make it compatible with the pattern precedence change.
Alongside this, I updated the documentation to include a `"**CHANGES**"` section when applicable, and use feature level 159 under 7.0 instead.

PIG208 commented:
The PR is rebased. But I need to also add the stubs for `url-template` because `linkifier.ts` is typed.

PIG208 commented:
I have updated the PR, and I think it is ready for integration review. It is mostly unchanged except for some renumberings of the migrations, versions and feature level.

I have also included type definitions for `url–template` since linkifier.ts is now typed. We are still using the older version, so the type definitions are ported from the latest one.

The PR is also updated to accomodate recent changes to the instruction for creating linkifiers and displaying `Linkifier` objects from using `str` to using `repr`.

PIG208 commented:
Since the last update, I fixed an issue with the puppeteer tests relevant to using the update linkifier modal with an invalid URL template, that accidentally fails the `clean()` check before the intended validator check. `RealmFilter.clean` was updated to return early when the URL template is invalid to begin with.

Regarding error handling for creating linkifiers, there is something that we can work on in the future.

- Instead of having validators defined on `url_template` field and `clean()` for `RealmFilter`, it might be better to just do the checks in `clean` and get rid of the validator. Because `clean()` relies on `url_template` being valid.
- Currently, we just raise a `ValiadtionError` saying that the "Invalid URL template.` without a more specific reason. We can try to look at the specific error being raised and give a more actionable error message.

For now, I don't want to expand the scope of this PR too much, so better error handling can be a separate task to follow up.

timabbott commented on requirements/common.in:
```diff
@@ -16,6 +16,7 @@ Markdown
 importlib-metadata ; python_version < "3.10"  # for Markdown
 Pygments
 jsx-lexer
+uri-template
```
Looks like you lost the `PROVISION_VERSION` bump for this.

timabbott commented on zerver/migrations/0434_backfill_realmfilter_url_template.py:
```diff
@@ -0,0 +1,48 @@
+                f'Failed to convert url format "{var_pattern}". The converted template "{converted_template}" is invalid.'
+            )
+        linkifier.url_template = converted_template
+    linkifier_model.objects.bulk_update(linkifiers, fields=["url_template"])
```
@alexmv @andersk can you help review this migration is correct?

timabbott commented on zerver/models.py:
```diff
@@ -1369,7 +1366,7 @@ def linkifiers_for_realm_remote_cache(realm_id: int) -> List[LinkifierDict]:
             LinkifierDict(
                 pattern=linkifier.pattern,
-                url_format=linkifier.url_format_string,
+                url_template=linkifier.url_template,
```
This is an API change that we'll need to manage. First, we need to discuss in "api design" how to avoid breaking the mobile apps with this transition -- one possibility might involve try to construct a backwards-compatibility `url_format` value to send to clients where possible, but it's very possible we'll need to do something fancier involving sending the empty list or something -- we don't want existing clients to crash due to the `url_format` field being missing.

Second, we need to document whatever exact API change we settle on in the API changelog and `zulip.yaml`.

timabbott replied:
Ahh, I see you have the docs in a later commit, but we still need to work out in "#api design" a compatibility plan to avoid breaking the mobile apps's markdown rendering.

timabbott commented on help/add-a-custom-linkifier.md:
```diff
@@ -31,28 +31,32 @@ This is best explained by example.
 
+Linkifiers use [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570.html)
+compliant URL templates to represent the link to be generated from the
+pattern.
```
This probably should appear a bit earlier in the page.

PIG208 replied:
Moved it to the intro part.

PIG208 commented:
The API change discussed [here](https://chat.zulip.org/#narrow/stream/378-api-design/topic/URI.20Template.20API) has been implemented in the commit `(squash) linkifiers: Update event queue APIs for compatibility.` We are adding a client capability which make event types `realm_filters` and `realm_linkifiers` return empty lists when not supplied with `true`.

It mainly changes the event queue API and the relevant documentation/tests. We do have some API changes to endpoints that add/update linkifiers from `linkifier: Support URL templates for linkifiers.`, whose documentation is also included in the commit mentioned.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -181,6 +184,7 @@ def from_dict(cls, d: MutableMapping[str, Any]) -> "ClientDescriptor":
             d.get("stream_typing_notifications", False),
             d.get("user_settings_object", False),
             d.get("pronouns_field_type_supported", True),
+            d.get("linkifier_url_template", False),
```
I'm not exactly sure where this needs to be, but I believe the `notify_linkifiers` code path will send new-format linkifier objects to clients that don't have the `linkifier_url_template` client capability. I believe the fix for this likely involves having any `realm_filters` events just silently dropped for clients who lack that, since the plan is we're just giving `[]` as the set of realm filters.

PIG208 replied:
OK, I got it updated so that we are dropping `realm_filters` unconditionally. For `realm_linkifiers`, `apply_event` will now check `linkifier_url_template` to decide if a state change is needed.

PIG208 commented:
I have rebased and updated the PR.

The original commit that implements the initial support for backwards-compatibility is mostly unchanged, except that the `zulip.yaml` change got ripped out.

The updates to the PR are organized into the follow-up commit that addresses the `notify_linkifiers` comment and the issues mentioned by Greg. The latter is mostly about further removing the moving parts of `realm_filters`, and wording updates to the API documentation.

alexmv commented on zerver/migrations/0439_backfill_realmfilter_url_template.py:
```diff
@@ -0,0 +1,48 @@
+            "}": "%7D",
+        }
+    )
+    for linkifier in linkifiers:
+        converted_template = linkifier.url_format_string.translate(escape_table)
+        # Replace format string variables with the RFC 6570 URI Template syntax
+        converted_template = var_pattern.sub(r"\1{\2}", converted_template).replace("%%", "%")
+        if not uri_template.validate(converted_template):
+            raise RuntimeError(
+                f'Failed to convert url format "{var_pattern}". The converted template "{converted_template}" is invalid.'
+            )
+        linkifier.url_template = converted_template
+    linkifier_model.objects.bulk_update(linkifiers, fields=["url_template"])
```
This looks good to me, and I've validated that it works with all patterns currently in Zulip Cloud.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -3109,6 +3109,15 @@ paths:
+                                    See [`POST /register`](/api/register-queue#parameter-client_capabilities).
+                                    for how client capabilities can be specified.
+
+                                    **Changes**: New in Zulip 7.0 (feature level 175). Linkifiers that
+                                    use [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570.html) compliant URL template
+                                    are supported.
```
This is saying that this property is new in Zulip 7.0.

Instead, it should say what changed. Because the new state of things is already described in the main text of the description (and of the descriptions of properties deeper in the tree), the emphasis should be on clearly describing the old state, doing so in terms of a diff from the new state.

Here, because the change from `url_format` to `url_template` is already covered below, I think the only difference that this particular "Changes" item needs to cover is that the bit about "will always receive an empty list unless" was new in this version.

PIG208 replied:
I added "will always receive an empty list unless" to the "changes" section of the event instead of the property.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -3118,11 +3127,14 @@ paths:
+                                          The [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570.html) compliant
+                                          URL template to be used for linkifying matches.
 
+                                          **Changes**: New in Zulip 7.0 (feature level 175). This replaced `url_format`
+                                          for linkification.
```
```suggestion
                                          **Changes**: New in Zulip 7.0 (feature level 175). This replaced `url_format`,
                                          which contained a URL format string.
```

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -3151,6 +3163,10 @@ paths:
                                 **Changes**: Deprecated in Zulip 4.0 (feature level 54), replaced by
                                 the `realm_linkifiers` event type, which has a clearer name and format,
                                 instead.
+
+                                **Changes**: In Zulip 7.0 (feature level 175), support for
+                                linkifiers in the legacy format was dropped by unconditionally
+                                returning an empty list at all times.
```
Newest change first:

```suggestion
                                **Changes**: In Zulip 7.0 (feature level 175), support for
                                linkifiers in the legacy format was dropped by unconditionally
                                returning an empty list at all times.

                                Deprecated in Zulip 4.0 (feature level 54), replaced by
                                the `realm_linkifiers` event type, which has a clearer name and format,
                                instead.
```

That's important so that when the reader sees a version number that's older than they care about, they can stop reading.

gnprice replied:
Also, this is a spot where the main description needs updating so that it reflects the current state. Which is that this event is never sent at all, right?

PIG208 replied:
That's right. This has been updated.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -3168,23 +3184,14 @@ paths:
                                         - type: integer
                                         - type: string
                                   description: |
-                                    An array of tuples, where each tuple describes a linkifier.
-                                    The first element of the tuple is a
-                                    string regex pattern which represents the pattern that should
-                                    be linkified on matching. The second element is the URL with which the
-                                    pattern matching string should be linkified with and the third element
-                                    is the ID of the realm filter.
+                                    An array that is always empty. It is only intended
+                                    for linkifiers created using the legacy format, which no longer exists
+                                    after being replaced by linkifiers with
+                                    [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570.html) compliant URL template.
```
Then once the description of the overall event makes clear that the current state is that this is never sent at all, I think this description can be left as it was.

The information in the existing description should be kept around in any case — it's needed for clients supporting older servers. The alternative to keeping it in the main text of the description would be to move it into a "Changes" item.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -9913,10 +9920,14 @@ paths:
+                                The [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570.html) compliant
+                                URL template to be used for linkifying matches.
+
+                                **Changes**: New in Zulip 7.0 (feature level 175). This replaced `url_format`
+                                for linkification.
```
```suggestion
                                **Changes**: New in Zulip 7.0 (feature level 175). This replaced `url_format`,
                                which contained a URL format string.
```

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -10274,6 +10285,14 @@ paths:
+
+            - `linkifier_url_template`: Boolean for whether the client accepts linkifiers
+              that uses [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570.html) compliant URL
+              template for linkification. If false, the client will get empty lists for event
+              types `realm_linkifiers`.
```
It's an empty list for `realm_linkifiers` in the register response, and no `realm_linkifiers` events at all, right?

PIG208 replied:
I'm looking for a better way to phrase this. Currently using "If false, an empty list will be sent in the `/register` response and `realm_linkifiers` events will be dropped".

gnprice replied:
That seems pretty good. I'd just add an explicit mention of which property in the `/register` response you're talking about.

Oh and as written, it sounds like the default is effectively *true*.

So combining those, one version would be:

> If false or unset, then the `realm_linkifiers` list in the `/register` response will be empty if present, and no `realm_linkifiers` events will be sent.

PIG208 replied:
OK, that's updated.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -18191,16 +18223,21 @@ components:
-        `%(name_of_the_capturing_group)s`.
+        `{name_of_the_capturing_group}`.
+
+        **Changes**: New in Zulip 7.0 (feature level 175). This replaced
+        `LinkifierURLFormatString`.
```
Is `LinkifierURLFormatString` a term that's meaningful to the reader of the API docs? I believe it's a name that's purely internal to the implementation of the API docs.

(And might have a role somewhere if generating API bindings from these OpenAPI definitions, too — though since we're not exercising that, there's no guarantee that it makes much sense in that context.)

If I'm understanding correctly the context of how this is used — which is as a parameter in the two routes to create or update a linkifier — then one version would be:

```suggestion
        **Changes**: New in Zulip 7.0 (feature level 175). This replaced
        the parameter `url_format_string`, which was a format string in which
        named groups' content could be inserted with `%(name_of_the_capturing_group)s`.
```

gnprice reviewed (commented):
Thanks for the revision!

Several comments below on the API docs. There's a question about the API behavior embedded in a couple of them: it sounds like when the client doesn't send the capability, we'll be sending `realm_linkifiers` events, just with an empty list of linkifiers. It'd be cleaner to avoid sending the events at all.

PIG208 commented on zerver/openapi/zulip.yaml:
```diff
@@ -10680,7 +10714,10 @@ paths:
 
-                          An array of tuples (fixed-length arrays) where each tuple describes
+                          An array that is always empty. It is only intended
+                          for linkifiers created using the legacy format.
```
I moved that old portion below "Changes" with the empty line removed.

PIG208 commented:
This seems to be failing for a network error.

timabbott commented:
I'm squashing the to-squash commit, renumbering migrations, and did the following bit of API docs copy-editing:

``` diff
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 380a27c023..e5bf21a175 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -3102,8 +3102,9 @@ paths:
                                 for how client capabilities can be specified.
 
                                 **Changes**: Changed in Zulip 7.0 (feature level 176). This event no longer
-                                gets sent by unless the `linkifier_url_template` client capability is set to
-                                `true`.
+                                gets sent unless the `linkifier_url_template` client capability is set to
+                                `true`, due to the backwards-incompatible change from specifying the
+                                URL from a `url_format` format string to `url_template`.
 
                                 **Changes**: New in Zulip 4.0 (feature level 54), replacing the
                                 previous `realm_filters` event type, which is still sent for
@@ -10725,7 +10726,7 @@ paths:
                           When present, this is always an empty array.
 
                           **Changes**: Became empty in Zulip 7.0 (feature level 176). In previous versions,
-                          There is an array of tuples (fixed-length arrays) where each tuple describes
+                          there is an array of tuples (fixed-length arrays) where each tuple describes
                           a single [linkifier](/help/add-a-custom-linkifier).
                           The first element of the tuple is a string regex pattern which represents
                           the pattern that should be linkified on matching.
```

timabbott commented:
I did another read through everything and this looks great; after renumbering the migrations following a different migration 0439 being merged, I've marked this to merge once CI passes. Thanks for all the work on this transition @PIG208!

PIG208 commented:
I greatly appreciate all the reviews!

andersk commented on web/src/linkifiers.ts:
```diff
@@ -73,20 +83,28 @@ function python_to_js_linkifier(pattern: string, url: string): [RegExp | null, s
     }
-    return [final_regex, url];
+    const url_template = url_template_lib.parse(url);
+    blueslip.info(`Linkifier info ${String(final_regex)} ${url}`, group_number_to_name);
```
Did we leave in this debug message intentionally?

PIG208 replied:
No. This needs to be removed.

PIG208 replied:
#25822
